keras_retinanet/losses.py

```diff
@@ -20,48 +20,62 @@
 
 def focal(alpha=0.25, gamma=2.0):
     def _focal(y_true, y_pred):
-        # discard batches, throw all labels / classifications on one big blob
-        labels         = keras.backend.reshape(y_true, (-1, keras.backend.shape(y_true)[2]))
-        classification = keras.backend.reshape(y_pred, (-1, keras.backend.shape(y_pred)[2]))
 
-        # filter out "ignore" anchors
-        anchor_state   = keras.backend.max(labels, axis=1)  # -1 for ignore, 0 for background, 1 for object
-        indices        = keras_retinanet.backend.where(keras.backend.not_equal(anchor_state, -1))
-        classification = keras_retinanet.backend.gather_nd(classification, indices)
-        labels         = keras_retinanet.backend.gather_nd(labels, indices)
-        anchor_state   = keras_retinanet.backend.gather_nd(anchor_state, indices)
+        # compute the divisor: for each image in the batch, we want the number of positive anchors
+
+        # override the -1 labels, since we treat -1/0 the same way for determining the divisor
+        divisor   = keras_retinanet.backend.where(y_true < 0, keras.backend.zeros_like(y_true), y_true)
+        divisor   = keras.backend.max(divisor, axis=2)
+        divisor   = keras.backend.cast(divisor, keras.backend.floatx())
+        # compute the number of positive anchors
+        divisor   = keras.backend.sum(divisor, 1)
+        #  ensure we do not divide by 0
+        divisor   = keras.backend.maximum(1.0, divisor)
+        # pad the tensor to have shape (batch_size, 1, 1) for future division
+        divisor   = keras.backend.expand_dims(divisor, axis=1)
```

Comment:
If you add `keepdims=True` to the `max` and `sum` functions, you don't need these 2 lines.